neqo-transport/src/recovery.rs

```diff
@@ -211,12 +228,102 @@ impl LossRecoverySpaces {
     }
 }
 
+#[derive(Debug)]
+struct CongestionControl {
+    congestion_window: u64, // = kInitialWindow
+    bytes_in_flight: u64,
+    congestion_recovery_start_time: Option<Instant>,
+    ssthresh: u64,
+}
+
+impl Default for CongestionControl {
+    fn default() -> Self {
+        CongestionControl {
+            congestion_window: INITIAL_WINDOW,
+            bytes_in_flight: 0,
+            congestion_recovery_start_time: None,
+            ssthresh: u64::max_value(),
+        }
+    }
+}
+
+impl CongestionControl {
+    fn on_ack_received(&mut self, pkt: &SentPacket) {
+        self.bytes_in_flight -= u64::try_from(pkt.size).unwrap();
+
+        if self.in_congestion_recovery(pkt.time_sent) {
+            // Do not increase congestion window in recovery period.
+            return;
+        }
+        if self.app_limited() {
+            // Do not increase congestion_window if application limited.
+            return;
+        }
+
+        if self.congestion_window < self.ssthresh {
+            // Slow start.
+            self.congestion_window += u64::try_from(pkt.size).unwrap();
+        } else {
+            // Congestion avoidance.
+            self.congestion_window += u64::try_from(MAX_DATAGRAM_SIZE * pkt.size as usize).unwrap()
+                / self.congestion_window
+        }
+    }
+
+    fn on_packets_lost(
+        &mut self,
+        now: Instant,
+        largest_acked_sent: Option<Instant>,
+        pto: Duration,
+        lost_packets: &[SentPacket],
+    ) {
+        for pkt in lost_packets {
+            self.bytes_in_flight -= u64::try_from(pkt.size).unwrap();
```

Comment:
checked_sub?

Reply:
this should never happen so let's try some asserts.